### PR TITLE
Added telemetries

### PR DESCRIPTION
**Pull request by kuvinodms**

### CODE REVIEW

File: Tasks/ServiceFabricDeployV1/ServiceFabricSDK/Publish-NewServiceFabricApplication.ps1 [hunk 1]
```diff
@@ -105,7 +105,11 @@
 
         [Parameter(ParameterSetName = "ApplicationParameterFilePath")]
         [Parameter(ParameterSetName = "ApplicationName")]
-        [Switch]$CompressPackage
+        [Switch]$CompressPackage,
+
+        [Parameter(ParameterSetName = "ApplicationParameterFilePath")]
+        [Parameter(ParameterSetName = "ApplicationName")]
+        [ref]$status
```
Comment (bishal-pdMSFT) [inline Tasks/ServiceFabricDeployV1/ServiceFabricSDK/Publish-NewServiceFabricApplication.ps1]: $status is good candidate global variable - then you won't need to pass along.

File: Tasks/ServiceFabricDeployV1/ServiceFabricSDK/Publish-NewServiceFabricApplication.ps1 [hunk 2]
```diff
@@ -142,6 +146,7 @@
 
     if (!$SkipPackageValidation)
     {
+        $status.value = "NewApp-TestingApplicationPackage"
         $packageValidationSuccess = (Test-ServiceFabricApplicationPackage $AppPkgPathToUse)
```
Comment (bishal-pdMSFT) [inline Tasks/ServiceFabricDeployV1/ServiceFabricSDK/Publish-NewServiceFabricApplication.ps1]: Let's remove redundant Test-. Keep it only in deploy.ps1

File: Tasks/ServiceFabricDeployV1/ServiceFabricSDK/Publish-NewServiceFabricApplication.ps1 [hunk 3]
```diff
@@ -250,13 +257,14 @@
...
             {
+                $status.value = "NewApp-UnregisteringRegisteredApplicationType"
                 Write-Host (Get-VstsLocString -Key SFSDK_UnregisteringExistingAppType -ArgumentList @($names.ApplicationTypeName, $names.ApplicationTypeVersion))
                 $reg | Unregister-ServiceFabricApplicationType -Force -TimeoutSec $UnregisterPackageTimeoutSec
-                $ApplicationTypeAlreadyRegistered = $false
                 if (!$?)
                 {
                     throw (Get-VstsLocString -Key SFSDK_UnableToUnregisterAppType)
                 }
+                $ApplicationTypeAlreadyRegistered = $false
```
Comment (bishal-pdMSFT) [inline Tasks/ServiceFabricDeployV1/ServiceFabricSDK/Publish-NewServiceFabricApplication.ps1]: why this change?

File: Tasks/ServiceFabricDeployV1/ServiceFabricSDK/Publish-NewServiceFabricApplication.ps1 [hunk 4]
```diff
@@ -230,6 +236,7 @@
                     # It will unregister the existing application's type and version even if its different from the application being created,
                     if ((Get-ServiceFabricApplication | Where-Object {$_.ApplicationTypeVersion -eq $($app.ApplicationTypeVersion) -and $_.ApplicationTypeName -eq $($app.ApplicationTypeName)}).Count -eq 0)
                     {
+                        $status.value = "NewApp-UnregisteringApplicationType"
```
Comment (bishal-pdMSFT) [inline Tasks/ServiceFabricDeployV1/ServiceFabricSDK/Publish-NewServiceFabricApplication.ps1]: Rather than settings similar status multiple places, you can move it to separate method and call it from both New/upgrade file. I don't think adding "NewApp-" or "UpgradeApp-" prefix is necessary.

File: Tasks/ServiceFabricDeployV1/deploy.ps1 [hunk 5]
```diff
@@ -176,16 +178,22 @@ try {
...
         $publishParameters['Action'] = "RegisterAndCreate"
         $publishParameters['OverwriteBehavior'] = Get-VstsInput -Name overwriteBehavior
+        $publishParameters['Status'] = [ref]$status
 
         Publish-NewServiceFabricApplication @publishParameters
     }
-} finally {
+}
+catch {
+    Write-Telemetry "Task_InternalError" "TaskStatus: $status | Exception: $_.Exception.GetType().FullName"
```
Comment (bishal-pdMSFT) [inline Tasks/ServiceFabricDeployV1/deploy.ps1]: Remove words "TaskStatus:" and "Exception:". We can use convention to have status and exception type separated by | without any space.

File: Tasks/ServiceFabricDeployV1/deploy.ps1 [hunk 6]
```diff
@@ -176,16 +178,22 @@ try {
...
+}
```
Comment (bishal-pdMSFT) [inline Tasks/ServiceFabricDeployV1/deploy.ps1]: Don't you need to re-throw?

File: Tasks/ServiceFabricDeployV1/deploy.ps1 (hunk 5 above)
Comment (kuvinodms) [reply Tasks/ServiceFabricDeployV1/deploy.ps1]: Used it for readability. If you don't see any use then lets remove it.

File: Tasks/ServiceFabricDeployV1/ServiceFabricSDK/Publish-NewServiceFabricApplication.ps1 (hunk 1 above)
Comment (kuvinodms) [reply Tasks/ServiceFabricDeployV1/ServiceFabricSDK/Publish-NewServiceFabricApplication.ps1]: Lets us PassByReference as its more controlled and always safer then to use Global variable

Comment (kuvinodms) [reply Tasks/ServiceFabricDeployV1/ServiceFabricSDK/Publish-NewServiceFabricApplication.ps1]: Passing an extra optional parameter (with reference) still seems fine to me as there is no extra memory usage

File: Tasks/ServiceFabricDeployV1/ServiceFabricSDK/Publish-NewServiceFabricApplication.ps1 (hunk 4 above)
Comment (kuvinodms) [reply Tasks/ServiceFabricDeployV1/ServiceFabricSDK/Publish-NewServiceFabricApplication.ps1]: Will raise a separate PR for the same

File: Tasks/ServiceFabricDeployV1/ServiceFabricSDK/Publish-NewServiceFabricApplication.ps1 (hunk 1 above)
Comment (bishal-pdMSFT) [reply Tasks/ServiceFabricDeployV1/ServiceFabricSDK/Publish-NewServiceFabricApplication.ps1]: In this case there does not seem to be usual problems related to global variables. Before execution a command will set status and that will always be the truth - we don't have multi-threading or parallelization in our flow.
I am in favor of global as code will be cleaner without need to pass $status everywhere.

File: Tasks/ServiceFabricDeployV1/ServiceFabricSDK/Publish-NewServiceFabricApplication.ps1 (hunk 2 above)
Comment (bishal-pdMSFT) [reply Tasks/ServiceFabricDeployV1/ServiceFabricSDK/Publish-NewServiceFabricApplication.ps1]: Please file an issue for that - otherwise we will miss it.

File: Tasks/ServiceFabricDeployV1/ServiceFabricSDK/Publish-NewServiceFabricApplication.ps1 (hunk 4 above)
Comment (bishal-pdMSFT) [reply Tasks/ServiceFabricDeployV1/ServiceFabricSDK/Publish-NewServiceFabricApplication.ps1]: Do we really need to know whether it is upgrade or new when lets say registering the package? This information seems to be redundant.
On the positive side we will be able to move code to common functions and re-use them.

File: Tasks/ServiceFabricDeployV1/deploy.ps1 (hunk 5 above)
Comment (bishal-pdMSFT) [reply Tasks/ServiceFabricDeployV1/deploy.ps1]: It is redundant. As we will never need to see full string. We will always query this data via kusto and it is easier to write query without these extra characters

File: Tasks/ServiceFabricDeployV1/ServiceFabricSDK/Publish-NewServiceFabricApplication.ps1 (hunk 2 above)
Comment (kuvinodms) [reply Tasks/ServiceFabricDeployV1/ServiceFabricSDK/Publish-NewServiceFabricApplication.ps1]: #7492

File: Tasks/ServiceFabricDeployV1/ServiceFabricSDK/Publish-NewServiceFabricApplication.ps1 (hunk 4 above)
Comment (kuvinodms) [reply Tasks/ServiceFabricDeployV1/ServiceFabricSDK/Publish-NewServiceFabricApplication.ps1]: Removing it in the hope that we don't need to know whether code failed in CreateApp or UpgradeApp.

File: Tasks/ServiceFabricDeployV1/ServiceFabricSDK/Publish-NewServiceFabricApplication.ps1 (hunk 1 above)
Comment (kuvinodms) [reply Tasks/ServiceFabricDeployV1/ServiceFabricSDK/Publish-NewServiceFabricApplication.ps1]: Using global variable now.

File: Tasks/ServiceFabricPowerShellV1/ServiceFabricPowerShell.ps1 [hunk 7]
```diff
@@ -29,10 +29,10 @@ try
 {
     # Initialize Service Fabric.
     Import-Module $PSScriptRoot\ps_modules\ServiceFabricHelpers
-    $connectedServiceEndpoint = Get-VstsEndpoint -Name $serviceConnectionName -Require
 
+    $global:operationId = $SF_Operations.Undefined
```
Comment (sachinma) [inline Tasks/ServiceFabricPowerShellV1/ServiceFabricPowerShell.ps1]: why undefined?

Comment (sachinma) [reply Tasks/ServiceFabricPowerShellV1/ServiceFabricPowerShell.ps1]: Shouldn't it be connect?

File: Tasks/ServiceFabricDeployV1/deploy.ps1 [hunk 8]
```diff
@@ -107,11 +111,13 @@ try {
 
         if (!$skipValidation)
         {
+            $global:operationId = $SF_Operations.TestApplicationPackage
```
Comment (sachinma) [inline Tasks/ServiceFabricDeployV1/deploy.ps1]: Should we add Get Application as operation as well

File: Tasks/Common/Deployment/TelemetryHelper/TelemetryHelper.psm1 [hunk 9]
```diff
@@ -1,38 +1,65 @@
...
+    $erroCodeMsg = $telemetryCodes[$codeKey]
+
+    ## If no error is passed mark it as not available
+    if ([string]::IsNullOrEmpty($errorMsg))
+    {
+        $errorMsg = "No error details available"
+    }
+    $erroCode = ('"{0}":{1}' -f $erroCodeMsg, $errorMsg)
+    ## Form errorcode as json string
+    $erroCode = '{' + $erroCode + '}'
```
Comment (sachinma) [inline Tasks/Common/Deployment/TelemetryHelper/TelemetryHelper.psm1]: variable name can be improved

Comment (sachinma) [inline Tasks/Common/Deployment/TelemetryHelper/TelemetryHelper.psm1]: This doesn't ensure json string if char escaping etc is required. Use native ways to genrate the json

File: Tasks/Common/Deployment/TelemetryHelper/TelemetryHelper.psm1 [hunk 10]
```diff
@@ -1,38 +1,65 @@
...
+    $exceptionData = ""
+    try
+    {
+        $src = $error.InvocationInfo.PSCommandPath + "|" + $error.InvocationInfo.ScriptLineNumber
+        $exceptionTypes = ""
+        $exception = $error.Exception
+        do
+        {
+            $exceptionTypes += $exception.GetType().FullName + ";"
+            $exception = $exception.InnerException
```
Comment (sachinma) [inline Tasks/Common/Deployment/TelemetryHelper/TelemetryHelper.psm1]: Consider handling AggregateException as well

File: Tasks/ServiceFabricPowerShellV1/ServiceFabricPowerShell.ps1 (hunk 7 above)
Comment (bishal-pdMSFT) [reply Tasks/ServiceFabricPowerShellV1/ServiceFabricPowerShell.ps1]: it is set to connect in Connect helper Connect-ServiceFabricClusterFromServiceEndpoint.ps1

File: Tasks/ServiceFabricDeployV1/deploy.ps1 (hunk 8 above)
Comment (bishal-pdMSFT) [reply Tasks/ServiceFabricDeployV1/deploy.ps1]: Vinod did not add it initially. I am not sure whether he had a reason. But I agree ideally we should add.

Comment (bishal-pdMSFT) [reply Tasks/ServiceFabricDeployV1/deploy.ps1]: done

Comment (bishal-pdMSFT) [reply Tasks/ServiceFabricDeployV1/deploy.ps1]: Added a few other missing as well

File: Tasks/Common/Deployment/TelemetryHelper/TelemetryHelper.psm1 (hunk 9 above)
Comment (bishal-pdMSFT) [reply Tasks/Common/Deployment/TelemetryHelper/TelemetryHelper.psm1]: done.
